torch/csrc/jit/codegen/cuda/lower_index_hoist.cpp

```diff
@@ -12,419 +12,261 @@ namespace cuda {
   }
-
-  TORCH_INTERNAL_ASSERT(
-      !used_loops_.empty(),
-      "No loop used for indexing found. ",
-      consumer_indexed_id->toString());
-
-  TORCH_INTERNAL_ASSERT(
-      consumer_leaf_ids.size() == used_loops_.size(),
-      "consumer_leaf_ids.size() = ",
-      consumer_leaf_ids.size(),
-      ", used_loops_.size() == ",
-      used_loops_.size(),
-      ", loops.size() == ",
-      loops.size());
+  return nullptr;
 }
 
-bool CommonIndexKey::operator==(const CommonIndexKey& other) const {
-  auto gpu_lower = GpuLower::current();
+} // namespace
 
-  if (concrete_indexed_id_ != other.concrete_indexed_id_) {
-    return false;
+Val* CommonIndexMap::hoistIndex(
+    Val* index,
+    const std::vector<kir::ForLoop*>& loops,
+    int64_t parent_pos,
+    bool always_insert) {
+  if (index->isA<TensorView>() || index->isA<kir::TensorIndex>()) {
+    // Current implementation of index hoisting does not has data flow analysis.
+    // It assumes that allocating all indices at the beginning of a for loop
+    // satisfies all data dependency. However, when the index is an expression
+    // of a tensor, this approach will fail. For this case we just give up
+    // hoisting by returning a nullptr.
+    return nullptr;
   }
 
-  if (used_loops_.size() != other.used_loops_.size()) {
-    return false;
+  auto def = index->definition();
+  if (def == nullptr || index->isConstScalar()) {
+    return index;
   }
 
-  // Check if both CommonIndexKeys use the same loops. If not, it's
-  // still valid to share the same hoisted index as long as: 1) each
-  // loop pair is mapped with the CA index map, and 2) they are not
-  // instantiated as actual loops.
-  for (const auto i : c10::irange(used_loops_.size())) {
-    auto lhs_loop = used_loops_.at(i);
-    auto rhs_loop = other.used_loops_.at(i);
-    if (lhs_loop == rhs_loop) {
-      continue;
-    }
-    if (gpu_lower->caMap()->areMapped(
-            lhs_loop->iter_domain(),
-            rhs_loop->iter_domain(),
-            IdMappingMode::EXACT) &&
-        lhs_loop->isTrivial() && rhs_loop->isTrivial()) {
-      continue;
-    }
-    return false;
+  auto my_pos = findOutermostPosWithSatisfiedDependency(index, loops);
+  auto my_loop = getLoopAtPos(loops, my_pos);
+
+  // Check if `index` is already computed. If yes, just reuse it and return.
+  auto existing_subexpr = eliminateCommonSubexpression(index, my_loop);
+  if (existing_subexpr != nullptr) {
+    return existing_subexpr;
   }
 
-  for (const auto i : c10::irange(loop_index_vals_.size())) {
-    auto lhs_index = loop_index_vals_.at(i);
-    auto rhs_index = other.loop_index_vals_.at(i);
-    if (lhs_index == rhs_index) {
-      continue;
+  // Recursively hoist all the producers of `index`
+  bool changed = false; // if any of the inputs is replaced by an existing val
+  std::vector<Val*> inputs;
+  for (auto input : def->inputs()) {
+    inputs.emplace_back(hoistIndex(input, loops, my_pos));
+    if (inputs.back() == nullptr) {
```

Comment:
Yes, your understanding is correct. Currently, if you have
```
index = (i1 + i2) + T0[0]
```
then the `i1 + i2` will be hoisted, but if you have
```
index = T0[0] +  (i1 + i2)
```
then the `i1 + i2` will not be hoisted. I will change this PR so that `i1 + i2` can be hoisted in both cases. Shouldn't be hard.

Reply:
done